src/Frame.cc

```diff
@@ -998,192 +1041,232 @@ bool Frame::FillSpatialProfileData(int region_id, CARTA::SpatialProfileData& spa
+
+        // add profiles
+        std::vector<float> profile;
+        bool write_lock(false);
 
-        // can no longer select stokes, so can use image cache or tile cache
+        // for each widget config with the same stokes setting
+        for (auto& config : point_regions_spatial_config.second) {
+            size_t start(config.start());
+            size_t end(config.end());
+            int mip(config.mip());
 
-        if (_loader->UseTileCache() &&
-            (mip < 2 || !_loader->HasMip(2))) { // Use tile cache to return full resolution data or prepare data for decimation
-            profile.resize(end - start);
+            if (!end) {
+                end = config.coordinate().back() == 'x' ? _width : _height;
+            }
 
-            if (config.coordinate() == "x") {
-                int tile_y = tile_index(y);
-                bool ignore_interrupt(_ignore_interrupt_X_mutex.try_lock());
+            int requested_start(start);
+            int requested_end(end);
+
+            int decimated_start(start);
+            int decimated_end(end);
+
+            // Round the endpoints if we're going to decimate
+            if (mip >= 2 && !_loader->HasMip(2)) {
+                // These values will be used to resize the decimated data
+                decimated_start = std::ceil((float)start / (mip * 2)) * 2;
+                decimated_end = std::ceil((float)end / (mip * 2)) * 2;
+                // These values will be used to fetch the data to decimate
+                start = decimated_start * mip;
+                end = decimated_end * mip;
+                end = config.coordinate().back() == 'x' ? std::min(end, _width) : std::min(end, _height);
+            }
 
-                for (int tile_x = tile_index(start); tile_x <= tile_index(end - 1); tile_x += TILE_SIZE) {
-                    auto key = TileCache::Key(tile_x, tile_y);
-                    // The cursor has moved outside this chunk row
-                    if (!ignore_interrupt && (tile_index(_cursor.y, CHUNK_SIZE) != TileCache::ChunkKey(key).y)) {
-                        return have_profile;
-                    }
-                    auto tile = _tile_cache.Get(key, _loader, _image_mutex);
-                    auto tile_width = tile_size(tile_x, _width);
-                    auto tile_height = tile_size(tile_y, _height);
-
-                    // copy contiguous row
-                    auto y_offset = tile->begin() + tile_width * (y - tile_y);
-                    auto tile_start = y_offset + max(start - tile_x, 0);
-                    auto tile_end = y_offset + min(end - tile_x, tile_width);
-                    auto profile_start = profile.begin() + max(tile_x - start, 0);
-                    std::copy(tile_start, tile_end, profile_start);
-                }
+            profile.clear();
+            bool have_profile(false);
 
-                have_profile = true;
+            if (_loader->UseTileCache() && (mip < 2 || !_loader->HasMip(2)) &&
```

Comment:
This feels confusing, with many `if` statements at the same level, testing combinations of mip and _loader->HasMip and is_current_stokes; most of these existed prior to this PR.  What do you think about refactoring this? (double check my logic is correct!)  I also think GetDownsampledRasterData can be used for any stokes since it has a `stokes` parameter (@confluence) ?
```
bool downsample(mip >= 2);

if (downsample && _loader->HasMip(2)) {
    // GetDownSampledRasterData with stokes in config
} else {
    if (downsample) {
        // Set decimated_start/end and update start/end
    }

    if (is_current_stokes) {
        if (_loader->UseTileCache()) {
            // Get profile from _tile_cache
        } else {
            // Get profile from _image_cache
        }
    } else {
        // GetSlicerData
    }

    if (have_profile && downsample) {
        // decimate profile
    }
}
```